include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -0,0 +1,2779 @@
+    variables is returned. This allows an internal procedure to access the
+    host's variables. It is up to lowering to determine the contract between
+    the host and the internal procedure.
+  }];
+
+  let arguments = (ins fir_BoxProcType:$val);
+
+  let results = (outs fir_ReferenceType);
+}
+
+def fir_BoxRankOp : fir_SimpleOneResultOp<"box_rank", [NoSideEffect]> {
+  let summary = "return the number of dimensions for the boxed value";
+
+  let description = [{
+    Return the rank of a value of `box` type.  If the value is scalar, the
+    rank is 0.
+
+      %57 = fir.box_rank %40 : (!fir.box<!fir.array<*:f64>>) -> i32
+      %58 = fir.box_rank %41 : (!fir.box<f64>) -> i32
+
+    The example `%57` shows how one would determine the rank of an array that
+    has deferred rank at runtime. This rank should be at least 1. In %58, the
+    descriptor may be either an array or a scalar, so the value is nonnegative.
+  }];
+
+  let arguments = (ins fir_BoxType:$val);
+
+  let results = (outs AnyIntegerType);
+}
+
+def fir_BoxTypeDescOp : fir_SimpleOneResultOp<"box_tdesc", [NoSideEffect]> {
+  let summary = "return the type descriptor for the boxed value";
+
+  let description = [{
+    Return the opaque type descriptor of a value of `box` type. A type
+    descriptor is an implementation defined value that fully describes a type
+    to the Fortran runtime.
+
+      %7 = fir.box_tdesc %41 : (!fir.box<f64>) -> !fir.tdesc<f64>
+  }];
+
+  let arguments = (ins fir_BoxType:$val);
+
+  let results = (outs fir_TypeDescType);
+}
+
+// Record and array type operations
+
+def fir_CoordinateOp : fir_Op<"coordinate_of", [NoSideEffect]> {
+  let summary = "Finds the coordinate (location) of a value in memory";
+
+  let description = [{
+    Compute the internal coordinate address starting from a boxed value or
+    unboxed memory reference. Returns a memory reference. When computing the
+    coordinate of an array element, the rank of the array must be known and
+    the number of indexing expressions must equal the rank of the array.
+
+    This operation will apply the access map from a boxed value implicitly.
+
+    Unlike LLVM's GEP instruction, one cannot stride over the outermost
+    reference; therefore, the leading 0 index must be omitted.
+
+      %i = ... : index
+      %h = ... : !fir.heap<!fir.array<?:f32>>
+      %p = fir.coordinate_of %h, %i : (!fir.heap<!fir.array<?:f32>>, index) -> !fir.ref<f32>
+
+    In the example, `%p` will be a pointer to the `%i`-th f32 value in the
+    array `%h`.
+  }];
+
+  let arguments = (ins AnyRefOrBox:$ref, Variadic<AnyCoordinateType>:$coor);
+
+  let results = (outs fir_ReferenceType);
+
+  let assemblyFormat = [{
+    operands attr-dict `:` functional-type(operands, results)
+  }];
+
+  let verifier = [{
+    // Recovering a LEN type parameter only makes sense from a boxed value
+    for (auto co : coor())
+      if (auto *s = co.getDefiningOp())
+        if (dyn_cast_or_null<LenParamIndexOp>(s)) {
```

Comment:
You've already checked `s` for null, so you just need dyn_cast. Alternatively, you could merge the two if checks and continue using dyn_cast_or_null.